daemon/src/file_asset_source.rs

```diff
@@ -394,233 +446,278 @@ impl FileAssetSource {
+                let mut num_processed = 0;
+                let mut metadata_changes = HashMap::new();
+                while num_processed < to_process {
+                    match rx.recv() {
+                        Ok((pair, maybe_result)) => {
+                            match maybe_result {
+                                // Successful import
+                                Ok(result) => {
+                                    let path = &pair
                                     .source
                                     .as_ref()
                                     .or_else(|| pair.meta.as_ref())
                                     .expect(
                                         "a successful import must have a source or meta FileState",
                                     )
                                     .path;
-                                self.ack_dirty_file_states(&mut txn, &pair)?;
-                                // TODO put import artifact in cache
-                                metadata_changes.insert(
-                                    path.clone(),
-                                    result.map(|r| r.0.source_metadata()).unwrap_or(None),
-                                );
+                                    self.ack_dirty_file_states(txn, &pair)?;
+                                    // TODO put import artifact in cache
+                                    metadata_changes.insert(
+                                        path.clone(),
+                                        result.map(|r| r.0.source_metadata()).unwrap_or(None),
+                                    );
+                                }
+                                Err(e) => error!("Error processing pair: {}", e),
                             }
-                            Err(e) => error!("Error processing pair: {}", e),
+                            num_processed += 1;
+                            import_iter.next();
+                        }
+                        _ => {
+                            break;
                         }
-                        num_processed += 1;
-                        import_iter.next();
-                    }
-                    _ => {
-                        break;
                     }
                 }
-            }
-            let mut change_batch = asset_hub::ChangeBatch::new();
-            self.process_metadata_changes(&mut txn, metadata_changes, &mut change_batch)?;
-            asset_metadata_changed = self.hub.add_changes(&mut txn, change_batch)?;
-            Ok(())
-        })?;
+                let mut change_batch = asset_hub::ChangeBatch::new();
+                self.process_metadata_changes(txn, metadata_changes, &mut change_batch);
+                asset_metadata_changed = self.hub.add_changes(txn, change_batch)?;
+                Ok(())
+            })
+            .unwrap_or_else(|err| error!("error while processing: {}", err));
```

Comment:
if it's still a TODO to handle errors in this big scope, panic on error for now, better than getting in some weird state.